### ISSUE TITLE
XcodeEditor does not handle Emoji in build phase names well

### ISSUE THREAD

**Issue by edwardmp** [appsquickly/XcodeEditor#78]
Hi,

Thanks so much for developing this. It was exactly what I needed. I noticed one issue, CocoaPods now uses a emoji in various build phase names. These ordinarily are just strings with a UTF8-Code in them. You could just remove those Emoji (see https://gist.github.com/cihancimen/4146056 on how to detect them).

Furthermore, I needed some functionality that is not present in the current CocoaPods release. Any idea when the next release will be?

Best regards,
Edward

**Comment by edwardmp** [comment]
@jasperblues 
Decided to investigate this. It appears the problem does not originate in your code. As I understand it, XCodeEditor will write XML output to the pbxproj file. Xcode then converts it to another format (PLIST?). When I close Xcode and the conversion does not take place, all is file. XcodeEditor correctly outputs the Emoji.

However, when Xcode is open and the conversion happens, the emoji are corrupted. I think this is a bug in Xcode, I'll try to file a bug report at Apple.

Correct XML output by XcodeEditor:
<img width="755" alt="schermafbeelding 2016-06-22 om 19 26 15" src="https://cloud.githubusercontent.com/assets/1686739/16277174/18fa9304-38b2-11e6-80b2-5fe79da05b19.png">

Corrupted conversion by Xcode:
<img width="693" alt="schermafbeelding 2016-06-22 om 19 46 22" src="https://cloud.githubusercontent.com/assets/1686739/16277191/249109dc-38b2-11e6-9312-e373db63d5a2.png">

Possibly related to http://openradar.appspot.com/18512876 and http://openradar.appspot.com/13827044

**Comment by edwardmp** [comment]
I filed a radar, copy listed [here](http://openradar.appspot.com/radar?id=5143591328415744).
I could reproduce the bug by converting the `pbxproj` file to XML using `plutil`. So this issue has nothing to do with XcodeEditor itself, although it is still important to resolve.

Before Apple fixes this, I suggest stripping emoji from build script names?

**Comment by jasperblues** [comment]
@edwardmp Thanks for the information. I noticed some other projects (eg CocoaPods) recently stopped using emoji in the project files, so that makes sense.

**Comment by edwardmp** [comment]
@jasperblues 
Yes indeed, found a few issues in the CocoaPods repo that might be related. Let's see how (or if) Apple responds. 

The major issue here by the way is not necessarily the emoji no longer working: I noticed my pbxproj grew to 200 mb and Xcode freezing because somehow the affected build phases would get added multiple times. For the latter I'm not sure whether this is because of a deeper issue in Xcode or in XcodeEditor.

Anyway, if I create a pull request for (at least temporarily) stripping emoji would you consider merging it in?

**Comment by jasperblues** [comment]
Yes, that sounds like an excellent temporary solution. 

Speaking of bugs. There's a few bug fixes on another branch that I did for a client using this project. I had to strip out generics as they were on an earlier platform. . . need to merge those back some time.

**Comment by edwardmp** [comment]
OK great, I'll see what I can make of it the coming days. Aside for this, your library has been working well for me, so thanks for creating it 👍

**Comment by jasperblues** [comment]
:) You're welcome buddy. Thanks for using and, if you feel like it,  contributing too.

**Comment by edwardmp** [comment]
@jasperblues 
I created a pull request. I found [this](https://github.com/woxtu/NSString-RemoveEmoji) neat NSString category which does the job well. But since you currently don't have any dependencies, I just copied these 2 files to Utils.

I also added tests for it. However, as mentioned in the pull request the changes are not written to the project file yet. Using the debugger I see that all emojis are stripped from XCBuildShellScript & XCBuildShellScriptDefinition. Maybe you need to mark them as dirty somewhere so that the changes will be written to file.

**Comment by edwardmp** [comment]
@jasperblues you merged in my stuff but did you see my question to you as the implementation was not complete yet.

**Comment by edwardmp** [comment]
Got a response from Apple today:

> There are no plans to address this based on the following:
> 
> We do not officially support the XML file format or workflows involving it.

Thus, it is important that we remove all emoji in all build scripts, but I'm still waiting for your input how to achieve this @jasperblues

**Comment by jasperblues** [comment]
@edwardmp sorry for the delay in reply! Would you be able to check in a broken test? (It doesn't have to have asserts that fail if that is hard - just eyeball the plist file). I could then make it pass for you.

**Comment by edwardmp** [comment]
@jasperblues 
Hi Jasper sorry didn't have time until now to respond.

The issue is that the updated build scripts (without emoji) are somehow not applied/saved properly. When the project is saved I see the old build script names as if nothing has changed. This may be in some part of the code that determines which stuff should get overwritten? Any clue? If not I can try to make a test case but this will take me some more time

**Comment by edwardmp** [comment]
Hi @jasperblues 
Did you see my previous message? I'd like to know since a project I develop is delayed solely by the emoji issue. Hope to hear from you soon. Thanks.

**Comment by jasperblues** [comment]
@edwardmp I'm snowed in with work at the moment, so won't be able to look at it for a while.
